hudi-common/src/main/java/org/apache/hudi/common/model/PartialOverwriteWithLatestAvroPayload.java

```diff
@@ -0,0 +1,133 @@
+
+  public PartialOverwriteWithLatestAvroPayload(Option<GenericRecord> record) {
+    super(record); // natural order
+  }
+
+  @Override
+  public Option<IndexedRecord> combineAndGetUpdateValue(IndexedRecord currentValue, Schema schema) throws IOException {
+    if (recordBytes.length == 0) {
+      return Option.empty();
+    }
+
+    GenericRecord incomingRecord = bytesToAvro(recordBytes, schema);
+    GenericRecord currentRecord = (GenericRecord) currentValue;
+    if (isDeleteRecord(incomingRecord)) {
+      return Option.empty();
+    }
+    return Option.of(overwriteWithNonNullValue(schema, currentRecord, incomingRecord));
+  }
+
+  @Override
+  public int compareTo(OverwriteWithLatestAvroPayload oldValue) {
+    return this.orderingVal.compareTo(oldValue.orderingVal);
+  }
+
+  @Override
+  public OverwriteWithLatestAvroPayload preCombine(OverwriteWithLatestAvroPayload oldValue, Properties properties, Schema schema) {
+    if (null == schema) {
+      // using default preCombine logic
+      return super.preCombine(oldValue);
+    }
+
+    try {
+      Option<IndexedRecord> incomingOption = this.getInsertValue(schema);
+      Option<IndexedRecord> oldRecordOption = oldValue.getInsertValue(schema);
+
+      if (incomingOption.isPresent() && oldRecordOption.isPresent()) {
+        boolean inComingRecordIsLatest = this.compareTo(oldValue) >= 0;
+        // ordering two records by ordering value
+        GenericRecord firstRecord = (GenericRecord) (inComingRecordIsLatest ? oldRecordOption.get() : incomingOption.get());
+        GenericRecord secondRecord = (GenericRecord) (inComingRecordIsLatest ? incomingOption.get() : oldRecordOption.get());
+        GenericRecord mergedRecord = overwriteWithNonNullValue(schema, firstRecord, secondRecord);
+        return new PartialOverwriteWithLatestAvroPayload(mergedRecord, inComingRecordIsLatest ? this.orderingVal : oldValue.orderingVal);
+      } else {
+        return super.preCombine(oldValue);
+      }
+    } catch (IOException e) {
+      return super.preCombine(oldValue);
+    }
+  }
+
+  private GenericRecord mergeRecord(Schema schema, GenericRecord first, GenericRecord second, BiFunction<Object, Object, Object> mergeFunc) {
+    schema.getFields().forEach(field -> {
+      Object firstValue = first.get(field.name());
+      Object secondValue = second.get(field.name());
+      first.put(field.name(), mergeFunc.apply(firstValue, secondValue));
+    });
+    return first;
+  }
+
+  /**
+   * Merge two records, the merged value of each filed will adopt the filed value from secondRecord if the value is not null, otherwise, adopt the filed value from firstRecord.
+   *
+   * @param schema record schema to loop fields
+   * @param firstRecord the base record need to be updated
+   * @param secondRecord the new record provide new field value
+   * @return merged records
+   */
+  private GenericRecord overwriteWithNonNullValue(Schema schema, GenericRecord firstRecord, GenericRecord secondRecord) {
```

Comment:
@stayrascal,

I really liked the idea of having a record payload that does partial merging. However, if I understood it correctly, what's proposed here is to do so in a very specific way: you're favoring the income record field's values, unless they are null (in which case, you would keep the existing one). I'm not saying this is not valuable, but that the idea of doing partial merging is so good that maybe we could have something more generic. I'm going to suggest a few changes in order to accomplish that:

- Make `PartialOverwriteWithLatestAvroPayload` an abstract class
- Instead of having `mergeFunc` as a parameter of the `mergeRecord` method, it could become an abstract method. This would lead to the removal of the `overwriteWithNonNullValue` method, which makes this implementation specific to your merging logic
- For the original use case (partial merge favoring non-null values), implement the proposed abstract class and implement the `mergeFunc` method with what you have in `overwriteWithNonNullValue`: `(first, second) -> Objects.isNull(second) ? first : second`

It's just an idea, that could make what you proposed useful for many more use cases. Hope this made sense, and thanks for bringing this idea!

Reply:
Hi @alvarolemos , thanks a lot for your useful suggestion. Yeah, I also considered to abstract the merge logic by using an abstract `merge` method or passing `merge` function into a generic function, and I choose the later. The reason as follow:
- the `preCombine` and `combineAndGetUpdateUpdate` might have different merge/combine logic, only implement one abstract `merge` function might not enough for both two cases. For example, these two methods in `OverwriteWithLatestAvroPayload` have different merge/combine logic.
- In current implementation, actually, the `mergeRecord` is a generic method even through it's a private method currently, but it don't care the detail merge logic and can be changed to protected/public scope if need. Instead, the `overwriteWithNonNullValue` is merge implementation in current "Payload", which is wrapper of `mergeFunc` and we can create two wrappers for `preCombine` and `combineAndGetUpdateValue` two scenarios if need, which is similar with what you mentioned about implement detail `mergeFunc` logic in sub class. We can still inherit this class implement detail `mergeFunc` logic, and pass to `mergeRecord` method.
- Another reason why i didn't chose creating abstract class currently is that there will only one sub class, we can refactor it if we have many case need to inherit this class, right now, just make it simple as much as possible.